TypeInfo.init -> .initializer (part 2)

aG0aep6G opened the pull request:
TypeInfo.init is deprecated and is going to to be removed.

This blocks <https://github.com/dlang/druntime/pull/1766>.

aG0aep6G committed:
TypeInfo.init -> .initializer

TypeInfo.init is deprecated and is going to to be removed.

aG0aep6G commented:
Regarding the tester failures: https://github.com/dlang/druntime/pull/1766#issuecomment-280845478

ibuclaw commented:
> Regarding the tester failures: dlang/druntime#1766 (comment)

Looks unrelated to the test failures to me.

schveiguy commented:
I know it's ugly, but perhaps:

```D
static if(__traits(compiles, {auto p = TypeInfo.initializer})) auto p = TypeInfo.initializer;
else auto p = TypeInfo.init
```

This is a very special and very odd case. I think we are justified in doing something like this in this one spot, even if it lasts forever!

One thing you could also do is create a UFCS function to create an initializer "property" when it doesn't already exist. Something like:

```D
static if(!__traits(compiles, ... /*same idea as above*/)
auto initializer(TypeInfo t)
{
   return t.init;
}
```

aG0aep6G commented:
> One thing you could also do is create a UFCS function to create an initializer "property" when it doesn't already exist.

Done. Figuring out why it segfaulted at first wasn't fun. I'd rather update the compilers in the testers than forever having to deal with bugs that have long been fixed.

schveiguy commented:
Looks good I think.

> I'd rather update the compilers in the testers than forever having to deal with bugs that have long been fixed.

Probably true, but having something that is robust like this is a step forward either way.

schveiguy commented on src/root/rmem.d:
```diff
@@ -189,32 +189,40 @@ else
+
+        // TypeInfo.initializer for compilers older than 2.070
+        static if(!__traits(hasMember, TypeInfo, "initializer"))
+        private const(void[]) initializer(T : TypeInfo)(const T t)
```
Is there a reason why this has to be a template? If it does have to be a template, you can probably leave the attributes off as well.

aG0aep6G replied:
> Is there a reason why this has to be a template?

Reason is in the commit message: It's a template because old ClassInfo uses a member `byte[] init;` instead of overriding the method.

It could also be an overload for ClassInfo, but that would be longer and the template way seems a bit more robust.

> If it does have to be a template, you can probably leave the attributes off as well.

I guess so. I'd leave them in, though. The function needs all those attributes in order to call TypeInfo.init. I prefer spelling them out in such cases. But I can remove them, of course, if you want.

schveiguy replied:
> Reason is in the commit message

Hah, didn't even notice that, as you have to hover over the message to see the whole thing. I love asking already exactly answered questions 😝 

Looking at history, looks like the `init` [change](https://github.com/dlang/druntime/pull/1205) for `ClassInfo` was fixed in April of 2015. Hard to believe the testers are using such an old compiler...

> But I can remove them, of course, if you want.

No worries, I don't have commit rights here anyway. I'll leave it up to the DMD maintainers to decide.

aG0aep6G committed:
TypeInfo.initializer for older compilers

It's a template because old ClassInfo uses a member `byte[] init;` instead
of overriding the method.

wilzbach commented:
Hmm this introduced an ICE (under fonky circumstances) in Vibe.d: https://issues.dlang.org/show_bug.cgi?id=17451
